python/cudf/cudf/core/dataframe.py

```diff
@@ -359,8 +359,35 @@ def _setitem_tuple_arg(self, key, value):
                 for i, col in enumerate(columns_df._column_names):
                     self._frame[col].loc[key[0]] = value_cols[i]
             else:
-                for col in columns_df._column_names:
-                    self._frame[col].loc[key[0]] = value
+                if isinstance(value, cudf.DataFrame):
+                    if value.shape != self._frame.loc[key[0]].shape:
+                        raise ValueError(
+                            f"shape mismatch: value array of shape "
+                            f"{value.shape} could not be broadcast "
+                            f"to indexing result of shape "
+                            f"{self._frame.loc[key[0]].shape}"
+                        )
+                    for col in self._frame._column_names:
+                        if col in value:
+                            self._frame._data[col][key[0]] = value.iloc[
+                                :, columns_df._column_names.index(col)
+                            ]
+                        else:
+                            self._frame._data[col][key[0]] = cudf.NA
```

Comment:
Last time when we coded this I really just focused on getting it working. Looking at this again we might be able to do better: the complexity of looking up from a tuple is proportional to the size of the tuple, so the complexity here is at least the `O(len(self._column_name * len(value._column_name))`. I suggest we should compute the set of overlapping and non-overlapping column names from value and do two pass assignment. I think 

```suggestion
                    value_columns_names = set(value._column_names)
                    column_names  = set(columns_df._column_names) # note we use columns_df here, which only have columns chosen by key[1]
                    columns_specified = column_names & value_columns_names
                    columns_unspecified = column_names - value_column_names
                    
                    for col in column_specified:
                        columns_df[col][key[0]] = value._data[col]
                    for col in column_unspecified:
                        columns_df[col][key[0]] = cudf.NA
```

Reply:
Actually I did some micro-benchmarking, with the assumption that values is roughly on the same width as self. Creating a set has some overhead and it seems like the cut-off point is about 10 items in the list, after that it's worth the overhead.
```python
In [23]: import timeit

In [24]: def method1(l1, l2):
    ...:     for x in l1:
    ...:         y = 42 if x in l2 else 88

In [25]: def method2(l1, l2):
    ...:     s2 = set(l2)
    ...:     for x in l1:
    ...:         y = 42 if x in s2 else 88

In [29]: for n in list(range(20)) + [100, 1000]:
    ...:     l1, l2 = list(range(n)), list(range(n))
    ...:     repeat = 1000
    ...:     t1 = timeit.timeit(lambda: method1(l1, l2), number=repeat) / repeat
    ...:     t2 = timeit.timeit(lambda: method2(l1, l2), number=repeat) / repeat
    ...:     print(f"size {n}: method1: {t1}s, method2: {t2}s")
size 0: method1: 2.982349833473563e-07s, method2: 4.21603093855083e-07s
size 1: method1: 4.109319997951388e-07s, method2: 5.577340489253402e-07s
size 2: method1: 4.776509013026952e-07s, method2: 6.118790479376912e-07s
size 3: method1: 5.614690016955137e-07s, method2: 6.809320766478777e-07s
size 4: method1: 6.629079580307007e-07s, method2: 7.527470588684082e-07s
size 5: method1: 7.572959875687956e-07s, method2: 8.628229843452573e-07s
size 6: method1: 8.757190080359578e-07s, method2: 9.588970569893717e-07s
size 7: method1: 9.862079750746488e-07s, method2: 1.0223720455542207e-06s
size 8: method1: 1.1484409915283322e-06s, method2: 1.097802072763443e-06s
size 9: method1: 1.2954400153830648e-06s, method2: 1.2035670224577188e-06s
size 10: method1: 1.4925100840628147e-06s, method2: 1.2536930153146386e-06s
size 11: method1: 1.6333690145984291e-06s, method2: 1.344060990959406e-06s
size 12: method1: 1.9053829601034522e-06s, method2: 1.4153749216347932e-06s
size 13: method1: 2.0433519966900347e-06s, method2: 1.4987780014052988e-06s
size 14: method1: 2.2338220151141284e-06s, method2: 1.5756030334159731e-06s
size 15: method1: 2.541176974773407e-06s, method2: 1.7017120262607931e-06s
size 16: method1: 2.9468770371749997e-06s, method2: 1.817883108742535e-06s
size 17: method1: 3.020999953150749e-06s, method2: 1.8007179023697973e-06s
size 18: method1: 3.275012946687639e-06s, method2: 1.9545200048014522e-06s
size 19: method1: 3.6183400079607964e-06s, method2: 2.4178719613701106e-06s
size 100: method1: 6.0095528955571353e-05s, method2: 8.978074067272246e-06s
size 1000: method1: 0.005277899098000489s, method2: 8.638103201519698e-05s
```